remove length check for new libicu versions

olegrok opened the pull request:
A bug (https://unicode-org.atlassian.net/browse/ICU-21802) was
fixed. So let's remove our workaround for versions that contain a
fix.

Follow-up #27

olegrok committed:
remove length check for new libicu versions

A bug (https://unicode-org.atlassian.net/browse/ICU-21802) was
fixed. So let's remove our workaround for versions that contain a
fix.

Follow-up #27

Totktonada commented on icu-date.lua:
```diff
@@ -191,6 +191,21 @@ function _M:format(format)
+-- Workaround for https://unicode-org.atlassian.net/browse/ICU-21802
+-- See also https://github.com/tarantool/icu-date/issues/27
+local version = ffi.new('UVersionInfo')
+local ok = pcall(function() ffi.C.u_getVersion(version) end)
```
Just curious: in what circumstances an error can be raised here?

olegrok replied:
When statically built Tarantool doesn't export u_getVersion symbol. E.g. it's relevant for all Tarantool EE versions before https://github.com/tarantool/tarantool-ee/commit/c254bb1889dc2f6aff65b5e11d08e669a9af11c9 
Just to avoid breaking something.

Totktonada replied:
Got it, thanks.

Totktonada commented on icu-date.lua:
```diff
@@ -191,6 +191,21 @@ function _M:format(format)
   return uchar_to_string(result, result_length)
```
It's time to add testing against different libicu versions?

I see the [icu4c-71_1-Ubuntu20.04-x64.tgz](https://github.com/unicode-org/icu/releases/download/release-71-1/icu4c-71_1-Ubuntu20.04-x64.tgz) archive on the [release page](https://github.com/unicode-org/icu/releases/tag/release-71-1), it should be suitable here.

olegrok replied:
I think yes. Since "version detection" logic appeared.

olegrok replied:
Pushed https://github.com/tarantool/icu-date/pull/31/commits/9485d3a204af4e07d0faf384c87e24c23ee5f5bc

However my approach requires tarantool to be built from source. Also libicu is built from source too. Probably I started to use `ldconfig` too late during testing.

Totktonada replied:
I don't mind building from source, it is definitely better than lack of such testing.

I would keep testing with distribution default icu and keep this testing fast (using setup-tarantool) to receive fast feedback during development. How about moving testing against different icu versions into its own workflow?

Totktonada replied:
I would also appreciate ideas how to speed up it if you have some. You can add TODO comments right into the workflow file.

olegrok replied:
> I don't mind building from source, it is definitely better than lack of such testing.

I found a way how to avoid building libicu. Your idea with [icu4c-71_1-Ubuntu20.04-x64.tgz](https://github.com/unicode-org/icu/releases/download/release-71-1/icu4c-71_1-Ubuntu20.04-x64.tgz) works. I just missed to use ldconfig.

> I would also appreciate ideas how to speed up it if you have some. You can add TODO comments right into the workflow file.

After previous message I even don't know. Maybe we could download Tarantool EE with different bundled libicu. But probably it's not completely fair way.

Totktonada replied:
> this patch manually build tarantool and libicu.

Forgotten in the commit message.

> After previous message I even don't know. Maybe we could download Tarantool EE with different bundled libicu. But probably it's not completely fair way.

Sure, it is irrelevant now.

Totktonada commented on icu-date/ffi_cdef.lua:
```diff
@@ -252,6 +252,8 @@ return function(ffi, icu_version_suffix)
       UCAL_DECEMBER,
       UCAL_UNDECIMBER
     } UCalendarMonths;
+    typedef uint8_t UVersionInfo[20];
```
> The binary form of a version on ICU APIs is an array of 4 uint8_t. 

https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/uversion_8h.html#a61caef96ba86d8e86a4031925ea44e94

Here we have 20 bytes and I don't understand from where this number comes.

olegrok replied:
It's my bad. I took U_MAX_VERSION_STRING_LENGTH value instead of U_MAX_VERSION_LENGTH. Fixed. Thanks.

Totktonada reviewed (approved):
I didn't test it, but the code looks okay (aside of a few comments and questions).

olegrok committed:
add testing over different libicu versions

This patch introduces new workflow to test icu-date module against
different libicu versions. It's not so easy task and to perform it
we should remove old version (libicu66 and libicu-dev), install
prebuit libicu library (fortunately, it's possible to just
download it). Still we should build Tarantool from source.
Since it costs a bit more time than default workflow it was
decided to keep fast_testing as is and implement this logic as
separate workflow.

Totktonada commented on .github/workflows/fast_testing.yml:
```diff
@@ -18,18 +18,40 @@ jobs:
+          git clone https://github.com/tarantool/tarantool.git
+          cd tarantool
+          git checkout ${{ matrix.tarantool }}
+          cmake -DENABLE_DIST=ON .
```
Nit: It is better to disable feedback daemon to don't pollute its statistics.

Totktonada replied:
OTOH, we have it in usual builds, so it would not change the picture much. Nevermind.

olegrok replied:
> Nit: It is better to disable feedback daemon to don't pollute its statistics.

No problem. Fixed.

Totktonada commented:
Feel free to push to `master` or ask me to proceed.
